src/clang-c-frontend/clang_c_convert.cpp

```diff
@@ -1444,7 +1447,50 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
     if(get_expr(*compound.getInitializer(), initializer))
       return true;
 
-    new_expr = initializer;
+    typet t = initializer.type();
+
+    /* A compound literals is an LValue that has associated static or automatic
+     * storage, depending on whether it appears in file or block scope.
+     * Therefore, in C, for instance a pointer to it can be taken and used
+     * within the same block without restrictions, e.g.
+     *
+     *   int *p = &(int){0}, x = 42;
+     *   memcpy(p, &x, sizeof(int));
+     *   assert(*p == 42);
+     *
+     * It has the same semantics as an non-anonymous variable of the
+     * corresponding type. Thus, we introduce a new symbol to represent it.
+     */
+
+    /* "cl" stands for "compound literal", but that is too long for possibly
+     * frequent copies of the symbol's name and id. */
+    symbolt &cl = anon_symbol.new_symbol(context, initializer.type(), "cl$");
+    /* mimic get_default_symbol() w/o setting id and name as that's already
+     * taken care of by anon_symbol.new_symbol() above */
+    cl.mode = "C";
+    cl.module = get_modulename_from_path(location.file().as_string());
+    cl.location = location;
+    cl.type = t;
+
+    cl.static_lifetime = compound.isFileScope();
+    cl.is_extern = false;
+    cl.file_local = true;
+    cl.value = initializer;
+
+    new_expr = symbol_expr(cl);
+
+    if(current_block)
+    {
+      code_declt decl(new_expr);
+      decl.operands().push_back(initializer);
+
+      current_block->operands().push_back(decl);
+    }
+    else
+    {
+      /* XXX fbrausse: do we need to record the declaration on the top-level? */
```

Comment:
Do we have a test case for that? I suggest we abort with a message.

Reply:
Yes, I've addded this test case:
```c
int *p = &(int){5};

int main()
{
	assert(p);
	assert(*p == 5);
	++*p;
	assert(*p == 6);
}
```

Reply:
Strange, with `--symex-trace`, `p` is assigned twice, even though the result is `VERIFICATION SUCCESSFUL`:
```
[...]
        p=&cl$1;

        cl$1=5;

        __ESBMC_num_total_threads=0;

        __ESBMC_num_threads_running=0;

        p=0;
[...]
```

Reply:
maybe because p is global, we default initialize it to zero?